aten/src/ATen/native/Indexing.cpp

```diff
@@ -226,34 +231,192 @@ static std::tuple<Tensor, Tensor> makeLinearIndex(Tensor self, TensorList orig)
+  strides.insert(strides.begin() + dims_before, replacement_shape.size(), 0);
+  return src.as_strided(shape, strides);
+}
+
+static Tensor reshape_indexer(const Tensor& index, int64_t dims_before, int64_t dims_after) {
+  auto orig_shape = index.sizes();
+  auto shape = DimVector();
+  shape.append(dims_before, 1);
+  shape.append(orig_shape.begin(), orig_shape.end());
+  shape.append(dims_after, 1);
+  return index.reshape(shape);
+}
+
+AdvancedIndex::AdvancedIndex(const Tensor& src, TensorList indices_list)
+{
+  int64_t element_size_bytes = src.type().elementSizeInBytes();
+  int dims_before = 0, dims_after = 0, dims_indexed = 0;
+  IntList replacement_shape;
+  for (size_t dim = 0; dim < indices_list.size(); dim++) {
+    if (!indices_list[dim].defined()) {
+      if (dims_indexed == 0) {
+        dims_before++;
+      } else {
+        dims_after++;
+      }
+    } else {
+      dims_indexed++;
+      replacement_shape = indices_list[dim].sizes();
+      indexed_sizes.push_back(src.size(dim));
+      indexed_strides.push_back(src.stride(dim) * element_size_bytes);
+    }
+  }
+
+  this->dims_before = dims_before;
+  this->dims_after = dims_after;
+  this->src = restride_src(src, dims_before, dims_indexed, replacement_shape);
+
+  for (auto& index : indices_list) {
+    if (index.defined()) {
+      indices.push_back(reshape_indexer(index, dims_before, dims_after));
+    }
+  }
 
-  Tensor src, linearIndex;
-  std::tie(src, linearIndex) = makeLinearIndex(self, indices);
-  return src.take(linearIndex);
+  // For CUDA tensors, force all index tensors to have the same striding to
+  // simplify the CUDA kernel.
+  if (indices.size() >= 2 && this->src.type().device_type() == kCUDA) {
+    if (!all_strides_match(indices)) {
+      for (size_t i = 0; i < indices.size(); i++) {
+        indices[i] = indices[i].contiguous();
+      }
+    }
+  }
 }
 
-Tensor index_put(const Tensor & self, TensorList indices, const Tensor & value) {
+static AdvancedIndex make_info(Tensor self, TensorList orig) {
+  checkIndexTensorTypes(orig);
+  // first expand ByteTensor (boolean masks) into 1 or more LongTensors
+  auto indices = expandByteTensors(self, orig);
+  // next broadcast all index tensors together
+  try {
+    indices = expand_outplace(indices);
+  } catch (std::exception& e) {
+    AT_ERROR("shape mismatch: indexing tensors could not be broadcast together"
+             " with shapes ", shapes_as_str(indices));
+  }
+  // add missing null Tensors so that it matches self.dim()
+  while (indices.size() < (size_t)self.dim()) {
+    indices.emplace_back();
+  }
+  // if the non-null indices are not all adjacent, transpose self and indices
+  // together so that they're adjacent at the front
+  if (!hasContiguousSubspace(indices)) {
+    std::tie(self, indices) = transposeToFront(self, indices);
+  }
+  return AdvancedIndex(self, indices);
+}
+
+static Tensor make_bogus_tensor(const Tensor& self, const AdvancedIndex& info) {
```

Comment:
I can't find any place where this function is used?
Does it have a special purpose?